### ISSUE TITLE
GPU Synchronization issue with .topk

### ISSUE THREAD

**Issue by trypag** [pytorch/pytorch#3396]
Hi,
I was able to reproduce a configuration in which I have what I believe is a GPU synchronization issue :
```python
import torch
from torch.autograd import Variable


def accuracy_2d(output, target, topk=(1,)):
    """
    Computes the precision@k for the specified values of k
    Considers output is : NxCxHxW and target is : NxHxW
    """
    maxk = max(topk)
    total_nelem = target.size(0) * target.size(1) * target.size(2)
    _, pred = output.topk(maxk, 1, True, True)
    correct = target.unsqueeze(1).expand(pred.size())
    correct = pred.eq(correct)

    res = []
    for k in topk:
        correct_k = correct[:, :k].contiguous().view(-1).float().sum(0)
        res.append(correct_k.mul_(100.0 / total_nelem))

    return res


class AverageMeter(object):
    """Computes and stores the average and current value"""
    def __init__(self):
        self.reset()

    def reset(self):
        self.val = 0
        self.avg = 0
        self.sum = 0
        self.count = 0

    def update(self, val, n=1):
        self.val = val
        self.sum += val * n
        self.count += n
        self.avg = self.sum / self.count


top1 = AverageMeter()
top3 = AverageMeter()

target = torch.LongTensor(16, 172, 172).cuda()
target_var = Variable(target)
pred = Variable(torch.FloatTensor(16, 135, 172, 172).cuda())

prec1, prec3 = accuracy_2d(pred.data, target, topk=(1, 3))
top1.update(prec1[0], 16)
top3.update(prec3[0], 16)
```

I noticed that removing the last two lines of code does not produce the expected behaviour, I suggest this is where the synchronization happens. Also, I think this is related to how `.topk` handles large kernels (cf  @soumith).

**Comment by killeent** [comment]
@trypag thanks for the repro code. Can you be a little more specific about the unexpected behavior you encounter when removing the last two lines?

**Comment by trypag** [comment]
Sure, I think we can agree this kind of code should not take more than 1 second to run on a gpu, instead of this running this full repro will take an indefinite time (at least more than 30 sec, did not wait longer).
When removing the last two lines it runs smoothly as expected by this kind of work (1 sec). Adding the last two lines creates what I think is a GPU synchronization problem related to the call of `.topk` and results in what may be a deadlock.
I hope you will be able to reproduce

**Comment by soumith** [comment]
the last two lines can also be replaced by a `torch.cuda.synchronize()` to have the same issue.
On my machine the script as given takes 11 seconds, and without the last two lines it takes 6 seconds.

What GPU are you using?

**Comment by trypag** [comment]
I don't have the computer now, I remember it's a quadro MXXXX with 4gb. Do you think it's a real software issue or just realistic due to the task and the hardware ?

**Comment by apaszke** [comment]
Are all other CUDA scripts running fine? Maybe it's CUDA JITting the code?

**Comment by trypag** [comment]
I never had any kind of issues before with this computer, always found a solution. I don't know what you mean by CUDA scripts ?

**Comment by soumith** [comment]
so, the topk CUDA kernel is notoriously slow in certain shapes / dimensions. There's no easy fix without implementing particular algorithms from research papers, from what i remember (I had a discussion on this with @wickedfoo a while ago). I think you are running into a combination of:
- bad sizes / dims for topk
- having a really small register set or number of SMs (4GB quadro card sounds like that).

**Comment by wickedfoo** [comment]
topk is slow when it is not run in the innermost dimension.

The way for us to fix this within Torch would be to transpose the tensor in temporary memory such that the selection dimension is innermost, run topk, then transpose the results back.

If there is a reasonably large pool of temporary memory now within Torch (say, 100s of MB), then this approach is worth taking. Suffering the synchronization hit by calling cudaMalloc to get the temporary memory is probably worth taking in this situation as well.

**Comment by colesbury** [comment]
Transposing only helps a little. Shouldn't be too hard to write a kernel for this sort of input (large batch size, small k, small array size).

**Comment by hellojialee** [comment]
@colesbury Any progress here? I face this issue recently. I search the top K elements in a big moderately big Tensor with shape (N, 17, 512, 512), which is 5 times slower than the case with shape (N, 17, 128, 128). Though the change of speed is reasonable. I just want to speed the topk operation.

**Comment by wickedfoo** [comment]
@hellojialee over what dimension are you selecting the top K, and how big is N (typically)?

Note that there is a 16x size difference between the two examples you provided.

**Comment by hellojialee** [comment]
Yes, I know. Thank you @wickedfoo.  I reshape the search dimension into the last dim. The run time seems right, but I just wonder if we could further speed up.
